### PR TITLE
feat(DTFS-7852): filter deals with amendment in progress only checker user

### PR DESCRIPTION
**Pull request by rkontogianni**
## Introduction :pencil2:
This PR is about to get all amendements with status Ready for Checker's Approval and filter allDeals in the application and notices page for Checker user only

## Resolution :heavy_check_mark:
- Add DEALIDS_CHECKERS_APPROVAL_AMENDMENTS in libs common
- Add getAllAmendments layer in portal to get all amendments with submitted statuses
- Add getDealIdsCheckersApprovalAmendments function to transform allAmendments into an array of all ids with amendment in progress.
- Add in filtersQuery the DEALIDS_CHECKERS_APPROVAL_AMENDMENTS key
- Update escapeOperators file for DEALIDS_CHECKERS_APPROVAL_AMENDMENTS key
- Add unit tests
## Miscellaneous :heavy_plus_sign:
List any additional fixes or improvements.

### CODE REVIEW

File: portal-api/src/v1/helpers/escapeOperators.js [hunk 1]
```diff
@@ -60,13 +61,19 @@ const recursivelyReplaceEscapeOperators = (filters, result = {}) => {
     // This handles when there is an array -- ie in the case of AND and OR
     if (Array.isArray(filters[key])) {
       const newKeyName = getArrayKeyOperatorName(key);
-      if (!result[newKeyName]) {
-        result[newKeyName] = [];
+
+      if (newKeyName === 'dealIdsWithAmendmentsInProgress') {
```
Comment (zainUKEF) [inline portal-api/src/v1/helpers/escapeOperators.js:65]: is there a const for `'dealIdsWithAmendmentsInProgress'`?

Comment (zainUKEF) [inline portal-api/src/v1/helpers/escapeOperators.js:65]: documentation too please just so easy to understand

File: portal-api/src/v1/helpers/escapeOperators.js [hunk 2]
```diff
@@ -60,13 +61,19 @@ const recursivelyReplaceEscapeOperators = (filters, result = {}) => {
...
+        result._id = { $in: filters[newKeyName].map((dealId) => new ObjectId(dealId)) };
+      } else {
+        if (!result[newKeyName]) {
+          result[newKeyName] = [];
+        }
+        filters[key].forEach((condition, i) => {
```
Comment (zainUKEF) [inline portal-api/src/v1/helpers/escapeOperators.js:71]: new line please

File: portal/server/api.test.js [hunk 3]
```diff
@@ -90,3 +90,5 @@ describe('api.loginWithSignInLink', () => {
     await expect(loginPromise).rejects.toEqual(error);
   });
 });
+
+describe('api.getAllAmendments', () => {});
```
Comment (zainUKEF) [inline portal/server/api.test.js:94]: tests are missing!

File: portal/server/controllers/dashboard/deals/deals-filters-query.js [hunk 4]
```diff
@@ -12,7 +12,7 @@ const keywordQuery = require('./deals-filters-keyword-query');
  * @example ( 'true', [ dealType: ['BSS/EWCS'] ], { _id: '123', firstName: 'Mock' } )
  * @returns { AND: [ { 'bank.id': '9'} ], OR: [{ dealType: 'BSS/EWCS' }] }
  */
-const dashboardDealsFiltersQuery = (filters, user) => {
+const dashboardDealsFiltersQuery = async (filters, user, userToken) => {
```
Comment (zainUKEF) [inline portal/server/controllers/dashboard/deals/deals-filters-query.js:15]: can the params documentation be updated

File: portal/server/controllers/dashboard/deals/deals-filters-query.js [hunk 5]
```diff
@@ -23,9 +23,12 @@ const dashboardDealsFiltersQuery = (filters, user) => {
...
   if (isChecker && !isMaker) {
-    query.AND.push({
-      [CONSTANTS.FIELD_NAMES.DEAL.STATUS]: CONSTANTS.STATUS.DEAL.READY_FOR_APPROVAL,
-    });
+    const dealIdsWithAmendmentsInProgress = await getDealIdsWithAmendmentInprogress(userToken);
+    const dealIdsExist = dealIdsWithAmendmentsInProgress && dealIdsWithAmendmentsInProgress.length > 0;
+    const orQuery = {
+      OR: [{ [CONSTANTS.FIELD_NAMES.DEAL.STATUS]: CONSTANTS.STATUS.DEAL.READY_FOR_APPROVAL }, ...(dealIdsExist ? [{ dealIdsWithAmendmentsInProgress }] : [])],
+    };
+    query.AND.push(orQuery);
```
Comment (zainUKEF) [inline portal/server/controllers/dashboard/deals/deals-filters-query.js:31]: new lines please

File: portal/server/helpers/getDealIdsWithAmendmentInProgress.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+const { PORTAL_AMENDMENT_INPROGRESS_STATUSES } = require('@ukef/dtfs2-common');
+const api = require('../api');
+
+/**
+ * Retrieves deal IDs that have amendments in progress.
+ * @returns {Promise<Array<string>>} A promise that resolves to an array of deal IDs with amendments in progress.
```
Comment (zainUKEF) [inline portal/server/helpers/getDealIdsWithAmendmentInProgress.js:6]: missing param

File: portal-api/src/v1/helpers/escapeOperators.js (hunk 1 above)
Comment (rkontogianni) [reply portal-api/src/v1/helpers/escapeOperators.js:65]: I added a const in libs common

File: portal/server/controllers/dashboard/deals/deals-filters-query.js [hunk 7]
```diff
@@ -23,10 +24,18 @@ const dashboardDealsFiltersQuery = (filters, user) => {
...
   if (isChecker && !isMaker) {
-    query.AND.push({
-      [CONSTANTS.FIELD_NAMES.DEAL.STATUS]: CONSTANTS.STATUS.DEAL.READY_FOR_APPROVAL,
-    });
+    const dealIdsWithAmendmentsInProgress = await getDealIdsWithAmendmentInprogress(userToken);
+    const dealIdsExist = dealIdsWithAmendmentsInProgress && dealIdsWithAmendmentsInProgress.length > 0;
+    const orQuery = {
+      OR: [
+        { [CONSTANTS.FIELD_NAMES.DEAL.STATUS]: CONSTANTS.STATUS.DEAL.READY_FOR_APPROVAL },
+        ...(dealIdsExist ? [{ [DEALIDS_AMENDMENTS_INPROGRESS]: dealIdsWithAmendmentsInProgress }] : []),
```
Comment (abhi-markan) [inline portal/server/controllers/dashboard/deals/deals-filters-query.js:32]: * Can this be a variable please.
* For my understanding is this stipulating `dealIdsWithAmendmentsInProgress: ObjectId(123)` as a secondary OR condition?

File: libs/common/src/constants/amendments.ts [hunk 8]
```diff
@@ -50,6 +50,8 @@ export const AMENDMENT_QUERIES = {
   LATEST: 'latest',
 } as const;
 
+export const DEALIDS_AMENDMENTS_INPROGRESS = 'dealIdsWithAmendmentsInProgress';
```
Comment (abhi-markan) [inline libs/common/src/constants/amendments.ts:53]: Is this the MongoDB document field?

Comment (rkontogianni) [reply libs/common/src/constants/amendments.ts:53]: @abhi-markan yes

File: portal/server/controllers/dashboard/deals/deals-filters-query.js (hunk 7 above)
Comment (rkontogianni) [reply portal/server/controllers/dashboard/deals/deals-filters-query.js:32]: @abhi-markan this is dealIdsWithAmendmentsInProgress: ['123', '456'] as a secondary OR condition and in the backend I transform this to _id: { $in:  [ObjectId('123'), ObjectId('456')]. The reason I am doing this is because I can't use ObjectId or $in in portal and pass them with axios to portal-api.

File: libs/common/src/constants/amendments.ts (hunk 8 above)
Comment (rkontogianni) [reply libs/common/src/constants/amendments.ts]: @abhi-markan @zainUKEF According with the ticket https://ukef-dtfs.atlassian.net/browse/DTFS2-7884 checker cannot view Further Maker's Input Required, I updated the name of this field to DEALIDS_CHECKERS_APPROVAL_AMENDMENTS in order to be clearer that these dealIds are only for deals that checker user can visit.

File: portal/server/helpers/getDealIdsCheckersApprovalAmendments.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+const { PORTAL_AMENDMENT_STATUS } = require('@ukef/dtfs2-common');
+const api = require('../api');
+
+/**
+ * Retrieves deal IDs that have amendments status ready for checker's approval.
+ * @param {string} userToken - A valid user token.
+ * @returns {Promise<Array<string>>} A promise that resolves to an array of deal IDs with amendments in progress.
+ */
+const getDealIdsCheckersApprovalAmendments = async (userToken) => {
```
Comment (abhi-markan) [inline portal/server/helpers/getDealIdsCheckersApprovalAmendments.js:9]:
```suggestion
const getCheckersApprovalAmendmentDealId = async (userToken) => {
```

Just thinking might be a better name, though not a blocker.

File: portal/server/helpers/getDealIdsCheckersApprovalAmendments.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  try {
+    const amendments = await api.getAllAmendments({ statuses: [PORTAL_AMENDMENT_STATUS.READY_FOR_CHECKERS_APPROVAL], userToken });
```
Comment (abhi-markan) [inline portal/server/helpers/getDealIdsCheckersApprovalAmendments.js:11]:
```suggestion
    const statuses = [PORTAL_AMENDMENT_STATUS.READY_FOR_CHECKERS_APPROVAL];

    const amendments = await api.getAllAmendments({ statuses, userToken });
```

File: portal/server/helpers/getDealIdsCheckersApprovalAmendments.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    const dealIds = amendments.map((amendment) => amendment.dealId);
+    return dealIds;
```
Comment (abhi-markan) [inline portal/server/helpers/getDealIdsCheckersApprovalAmendments.js:13]:
```suggestion

    return dealIds;
```

File: portal/server/helpers/getDealIdsCheckersApprovalAmendments.test.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+const userToken = 'test-token';
+const mockAmendmentsCheckersApproval = [{ dealId: 'deal1' }, { dealId: 'deal2' }, { dealId: 'deal3' }];
+console.error = jest.fn();
+
+describe('getDealIdsCheckersApprovalAmendments', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('should return deal IDs when amendments are provided', async () => {
```
Comment (abhi-markan) [inline portal/server/helpers/getDealIdsCheckersApprovalAmendments.test.js:15]: Can we please add an assertion for an edge case.

```
it('should return an empty array when an undefined response is received from the API call', async () => {
    api.getAllAmendments.mockResolvedValue(undefined);

    const result = await getDealIdsCheckersApprovalAmendments(userToken);

    expect(result).toEqual([]);
  });
```

File: libs/common/src/constants/amendments.ts [hunk 13]
```diff
@@ -50,6 +50,8 @@ export const AMENDMENT_QUERIES = {
   LATEST: 'latest',
 } as const;
 
+export const DEALIDS_CHECKERS_APPROVAL_AMENDMENTS = 'dealIdsCheckersApprovalAmendments';
```
Comment (abhi-markan) [inline libs/common/src/constants/amendments.ts:53]:
```suggestion
export const CHECKERS_AMENDMENTS_DEAL_ID = 'checkersAmendmentDealId';
```

what are yout thoughts on above please?

File: portal/server/helpers/getDealIdsWithAmendmentInProgress.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Retrieves deal IDs that have amendments in progress.
+ * @param {string} userToken - A valid user token.
+ * @returns {Promise<Array<string>>} A promise that resolves to an array of deal IDs with amendments in progress.
+ */
+const getDealIdsWithAmendmentInprogress = async (userToken) => {
+  try {
+    const amendments = await api.getAllAmendments({ statuses: PORTAL_AMENDMENT_INPROGRESS_STATUSES, userToken });
+    const dealIds = amendments.map((amendment) => amendment.dealId);
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply portal/server/helpers/getDealIdsWithAmendmentInProgress.js]: Apologies Rena I missed it can it be `dealId` as we can also one single dealId.

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: portal/server/helpers/getDealIdsWithAmendmentInProgress.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (rkontogianni) [reply portal/server/helpers/getDealIdsWithAmendmentInProgress.js]: @abhi-markan this is an array of dealIds why to make it single?

File: portal/server/helpers/getDealIdsCheckersApprovalAmendments.js (hunk 9 above)
Comment (rkontogianni) [reply portal/server/helpers/getDealIdsCheckersApprovalAmendments.js]: I changed it to  getCheckersApprovalAmendmentDealIds because is an array of dealIds

**Review by abhi-markan — APPROVED** [review]

**Review by abhi-markan — APPROVED** [review]
